### ISSUE TITLE
Document thread safety

### ISSUE THREAD

**Issue by ioquatix** [rack/rack#1617]
I would like to start a discussion about thread safety of rack applications and defining a thread safety specification for rack 3.

Personally, I think the safest option is to define rack apps as unsafe to invoke from multiple threads. I can elaborate, but I'd be interested to hear other opinions first.

**Comment by jeremyevans** [comment]
I don't think the rack specification should concern itself with thread safety.  The rack API itself is functional (does not require side effects).  Why would `proc{|env| [200, {}, []]}` be unsafe to invoke from multiple threads?  Defining all rack apps as thread-unsafe just because some rack apps are thread-unsafe seems wrong to me.

**Comment by ioquatix** [comment]
> I don't think the rack specification should concern itself with thread safety.

I believe that too.

But the default assumption should be that methods are thread unsafe.

It's trivial to make a middleware that is not thread safe.

> Defining all rack apps as thread-unsafe just because some rack apps are thread-unsafe seems wrong to me.

"Defining all rack apps as thread-safe just because some rack apps are thread-safe seems wrong to me".

Do you believe a method is thread-safe or thread-unsafe by default? Because in my experience there is plenty of code which is the latter, and even some code which tries to be the former fails.

**Comment by jeremyevans** [comment]
Whether a rack app is thread safe or not is a property of the app, it has nothing to do with the rack specification or library, IMO.  I don't think rack should document whether rack apps are thread-safe or not, since rack has no control over rack apps.  I suppose the furthest I would be willing to go would be to say that "rack apps may not be thread-safe", or "whether a rack app is thread-safe depends on the app".  However, while true, I don't see the point in documenting that.

It is true that it is trivial to make a thread-unsafe middleware.  However, I would respond that it is even more trivial to write a thread-safe middleware: `class M; def initialize(a) @a = a end; def call(h) @a.call(h) end; end`

**Comment by ioquatix** [comment]
> I would respond that it is even more trivial to write a thread-safe middleware.

Thread safety is non-trivial when considering actual programs. A nice summary of different issues you might encounter: https://bearmetal.eu/theden/how-do-i-know-whether-my-rails-app-is-thread-safe-or-not/

> Whether a rack app is thread safe or not is a property of the app, it has nothing to do with the rack specification or library, IMO

Reentrancy is an important quality which should be documented. Either we expect rack apps to be thread safe, or we don't, in which case we can't assume any rack apps are thread safe in general.

The point of rack is to bridge web applications and servers. Either:

- The server states "I will invoke you from multiple threads" and the app crashes if that's not the case (what we have now).
- The server asks "I can invoke you from multiple threads, are you thread safe?" and loads the app once for multiple threads, or loads the app several times, once per thread.

Here are the options:

- Do nothing. Apps which are not thread safe continue to crash. [Bad for users](https://github.com/puma/puma/issues/2001).
- Document that rack apps are not thread safe by default. Servers should default to instantiating one app per thread. Document optional behaviour `--thread-safe` which allows a single instance per thread pool if applicable (opt in by user).
- Document that rack apps must be thread safe. Existing servers work as expected, users are expected to fix application code and build thread safety by default. This overhead affects servers which are multi-process and don't need thread safety.
- Allow servers to advertise thread safety in a standard way (e.g. injecting into builder a set of options like previously discussed). Apps can adjust internally to suit.
- Allow applications to communicate to the server that they are thread safe or not. It needs to happen before `config.ru` is loaded. Select appropriate default for thread safe or not.

cc @tenderlove

**Comment by nateberkopec** [comment]
> Defining all rack apps as thread-unsafe just because some rack apps are thread-unsafe seems wrong to me.

Agreed. Why does Rack have to concern itself with this issue? If people are writing thread-unsafe applications and then using them with Puma, isn't that Puma's problem? 

Rack is a protocol, not a set of best-practices for application development. Where does the remit come from that Rack has to make statements about the way the application is implemented?

> I don't think rack should document whether rack apps are thread-safe or not, since rack has no control over rack apps.

Perhaps we should declare that all Rack applications should be bug-free, as that would be better for users. A Rack application that has bugs should not be compliant with the Rack specification. :trollface:

**Comment by jeremyevans** [comment]
What we have now is the server tells the application whether it may invoke the application from multiple threads (`rack.multithread`), and it is up to the application to handle that.  Granted, by it being an entry in the rack environment, it comes too late to be generally useful.

I don't think Option 2 will work.  `run` can be passed a constant, which would result in the same object no matter how many builders you use.

Option 3 you can't enforce, so there is little point documenting it.

I think it will be simpler to adjust server behavior based on whether the application says it is thread-safe, compared to adjusting application behavior based on whether the server says it is multi-threaded.  So I guess option 5 before your edit made the most sense to me.  After your edit option 5 seems similar to option 2, and I don't think it would work.

Personally I'm fine with option 1.  I don't think it is bad for users, as least not bad for most users.  It may be bad for users who run thread-unsafe apps with threaded webservers, but I don't see that as a significant problem.

> If people are writing thread-unsafe applications and then using them with Puma, isn't that Puma's problem?

I would say it is the app developers problem, not Puma's problem.  But I agree it isn't Rack's problem.

**Comment by nateberkopec** [comment]
> adjust server behavior based on whether the application says it is thread-safe

From my perspective, you say that your application is thread-safe the moment you configure your server for an app server with thread count > 1. I'm not sure what an extra option adds here. Who out there would be setting their app server's thread count to 5 and then also selecting a "non-threadsafe" option in addition to that.

**Comment by ioquatix** [comment]
> But I agree it isn't Rack's problem.

Then why do we have `rack.multithread`? Can we remove it? As you say: "it comes too late to be generally useful."

> From my perspective, you say that your application is thread-safe the moment you configure your server for an app server with thread count > 1.

Why? It's perfectly reasonable on JRuby to use isolated threads for parallelism. This doesn't mean you want shared state... it's perfectly okay to make one app per thread.

> Perhaps we should declare that all Rack applications should be bug-free, as that would be better for users. A Rack application that has bugs should not be compliant with the Rack specification.

The problem here is that whether an app is thread safe or not, can be handled by the server. Bug or bug free is not something that can be easily mitigated by the server.

In the case that an app is not thread safe, the server can make one app instance per thread. In order to do that, the app needs to explicitly state somehow: "I am thread safe" or "I am not thread safe". Then, we need to decide what should be the default. It sounds like @jeremyevans is somewhat on board with this (option 5), I'd say it's probably the best option, but we still needed to define, what is the default - thread safe or not thread safe - or do we just expect servers to do whatever they are already doing?

Then, if we do choose that option, what about servers like unicorn or falcon multi-process - they don't need to incur the overhead of thread safety, so I'd suggest the server should say: "Your application is being loaded into a multi-process context, so thread safety is not required.".

Without changes within Rack, none of this is possible, so it's definitely a concern of Rack.

**Comment by jeremyevans** [comment]
> From my perspective, you say that your application is thread-safe the moment you configure your server for an app server with thread count > 1.

Agreed.

> Then why do we have `rack.multithread`? Can we remove it? As you say: "it comes too late to be generally useful."

I'm OK with removing that and `rack.multiprocess` in Rack 3.

> Then, we need to decide what should be the default. It sounds like @jeremyevans is somewhat on board with this (option 5)

I was on board with option 5 back when I thought it meant something else (before your edit).  I don't think you can instantiate separate apps per thread.  Apps can be constants in Ruby (e.g. `run Roda`).  At least doing so doesn't necessarily make the resulting apps thread-safe.

I don't think Rack needs to change or be involved.  Applications that are thread-safe can run with a multi-threaded server, and applications which are not can run with a single threaded server.  If users do not know whether their applications are thread-safe, they should probably run single threaded.

**Comment by ioquatix** [comment]
> I don't think Rack needs to change or be involved. Applications that are thread-safe can run with a multi-threaded server, and applications which are not can run with a single threaded server. If users do not know whether their applications are thread-safe, they should probably run single threaded.

The third option is, apps which are not thread safe can be run in isolation on separate threads. Is there a reason why we should ignore that?

**Comment by ioquatix** [comment]
So, after looking through the code changes required for #1618, I found that the session code is *still* racey:

https://github.com/rack/rack/blob/6196377654b7ff7ce7abaecea62bb285d77d53aa/lib/rack/session/pool.rb#L39-L44

It:

- Assumes that there is a shared `@pool` which isn't always true (multi-process, multi-thread with unique app, guild, etc).
- `generate_sid` is invoked from outside of the mutex context, creating race conditions.

So, not even rack itself is thread safe.

**Comment by boazsegev** [comment]
@ioquatix ,

> The third option is, apps which are not thread safe can be run in isolation on separate threads. Is there a reason why we should ignore that?

This option doesn't exist in practice - not with Ruby MRI (C Ruby).

It could also result in unexpected behavior.

Consider the following `config.ru`:

```ruby
# config.ru
require 'a_lot_of_stuff.rb'
APP = Proc.new {|e| [200, {"Content-Length" => "11" }, ["Hello World"]] }
run APP
```

APP is a global object - how could a server safely load this `config.ru` file with a different `APP` object per thread?

Worst - consider this:

```ruby
# config.ru
$my_lock = Motex.new
$counter = 0;
APP = Proc.new {|e| 
   number = $my_lock.synchronize { $counter += 1 }
   out = "Hello World #{number}"
   [200, {"Content-Length" => out.length.to_s }, [out]]
}
run APP
```

If you could load a different APP per thread, you'd get really funny results...

**Comment by ioquatix** [comment]
@boazsegev you've really hit the nail on the head. I completely agree with you it's a bad situation - what SHOULD we be doing with such a `config.ru` file?

Is this something servers should be expected to handle? How about when guilds are introduced? If users insist on such code, how should it work with JRuby where users might want isolated parallelism to mimic multi-process models?

The only way we can start to address this situation is to document *what we expect* w.r.t. an app. If your `config.ru` is setting constants and modifying global state, it's definitely going to be at odds with most execution models.

So, I personally believe we should document that such a design is ill-formed. The alternative is the wild-west you show in your examples, and application servers have no way to reason about what is safe.

**Comment by ioquatix** [comment]
@nateberkopec So, I checked your assumptions, and this is just a single sample, but with a fresh Rails app:

```
# One app for all threads:
{:TOTAL=>447127, :FREE=>79035, :T_OBJECT=>11914, :T_CLASS=>6174, :T_MODULE=>959, :T_FLOAT=>9, :T_STRING=>134890, :T_REGEXP=>1738, :T_ARRAY=>75667, :T_HASH=>56955, :T_STRUCT=>492, :T_BIGNUM=>13, :T_FILE=>64, :T_DATA=>4145, :T_MATCH=>14, :T_COMPLEX=>1, :T_RATIONAL=>756, :T_SYMBOL=>481, :T_IMEMO=>72409, :T_ICLASS=>1411}

# One app per thread:
{:TOTAL=>449997, :FREE=>83199, :T_OBJECT=>11503, :T_CLASS=>6174, :T_MODULE=>959, :T_FLOAT=>9, :T_STRING=>134449, :T_REGEXP=>1738, :T_ARRAY=>75324, :T_HASH=>56910, :T_STRUCT=>492, :T_BIGNUM=>13, :T_FILE=>64, :T_DATA=>4111, :T_MATCH=>14, :T_COMPLEX=>1, :T_RATIONAL=>756, :T_SYMBOL=>481, :T_IMEMO=>72389, :T_ICLASS=>1411}
```

So, it didn't create a huge number of extra objects.

That being said, it was a very basic app, and of course shared caches and other resources might impact memory usage. But at least it's not an extra 100,000 objects per thread or something obscene.

**Comment by jeremyevans** [comment]
So maybe we just document: "If you are not sure app is thread-safe, you should not run in a multi-threaded mode."  I definitely think we should not go further than that, and I don't think going that far is necessary.  Let's leave things up to the person running the application to choose an appropriate server.  If they choose wrongly, that's on them.  There's no need to add more complexity to Rack here, and doing so is as likely to make things worse for some as it is to make them better for others.

As I mentioned and as @boazsegev provided a code example for, you can't load `config.ru` multiple times in the same process to get separate values of the app per thread.  It could work on apps specifically designed to support such a configuration, but I'm guessing the vast majority of current rack apps cannot.  Not that it cannot work, but that doing so doesn't actually solve thread-safety issues.

As far as I know, Rails cannot support it, as you cannot run multiple Rails apps in the same process last I checked (not sure if that has changed in the last 6 or so years).  What code are you using to generate the Rails-app-per-thread results?

**Comment by ioquatix** [comment]
> As far as I know, Rails cannot support it, as you cannot run multiple Rails apps in the same process last I checked (not sure if that has changed in the last 6 or so years). What code are you using to generate the Rails-app-per-thread results?

I used falcon, and it worked fine, I didn't see any obvious issues. Can you explain why it shouldn't work? What did you check to verify that it doesn't work?

> Not that it cannot work, but that doing so doesn't actually solve thread-safety issues.

If the app returned from evaluating `config.ru` is unique, and the app server does not use that app object on multiple threads, it absolutely does solve the thread safety issue, at least as far as the relationship between rack, the server and the app is concerned. Internal to the app, obviously they can do whatever they want (bugs as described by @nateberkopec :rofl:).

**Comment by jeremyevans** [comment]
> If the app returned from evaluating `config.ru` is unique

This is not true for a large portion of Rack apps.  It's probably not true for almost all apps where the rack app is a class and not an instance.

> > As far as I know, Rails cannot support it, as you cannot run multiple Rails apps in the same process last I checked (not sure if that has changed in the last 6 or so years). What code are you using to generate the Rails-app-per-thread results?
> 
> I used falcon, and it worked fine, I didn't see any obvious issues. Can you explain why it shouldn't work? What did you check to verify that it doesn't work?

`Rails.application` is a singleton: https://github.com/rails/rails/blob/master/railties/lib/rails.rb#L39-L41

If you have multiple separate Rails applications in the same process, `Rails.application` cannot be correct for both.

Even if you were able to get app-per-thread working for the app itself, you would have to ensure all everything inside the application was duplicated, which again, will be true for very few apps.

Here's an example app giving you a different application object if config.ru is loaded multiple times, but probably a segfault if you run multiple apps in separate threads in the same process.

```ruby
# config.ru
require 'sequel'

# Turn on single threaded mode
Sequel.single_threaded = true
DB = Sequel.sqlite('file.sqlite3')

run(proc{|env| [200, {}, [DB[:table].get(:column)]]})
```

Basically, app-per-thread doesn't work as a concept in general.  Apps would have to be specifically designed for it.  If you are going to put in that work, it's probably better to just make the app thread-safe.

**Comment by ioquatix** [comment]
> Rails.application is a singleton: https://github.com/rails/rails/blob/master/railties/lib/rails.rb#L39-L41.If you have multiple separate Rails applications in the same process, Rails.application cannot be correct for both.

That's really an issue with the design of Rails, not Rack, though. I mean, sure, it might not work (but it seems to be fine in my testing). As long as the application is thread safe, I don't think it's an issue that `config.ru` returns an app that points at the same global application instance, i.e.

```
# config.ru
require 'app'

run APP
```

To me, this should be fine. It might not support graceful reloading in JRuby, but that's on the author of the application, not on the Rack specification.

> Even if you were able to get app-per-thread working for the app itself, you would have to ensure all everything inside the application was duplicated, which again, will be true for very few apps.

That's not true. If you want to use a global shared cache, you need to be sure it's thread safe, and that's okay.

Also, I tried your example, and it seems to work fine. Althought I get what you are trying to show. In addition, Ruby gives tons of warnings:

```
koyoko% falcon serve --threaded
 0.14s     info: Falcon::Command::Serve [oid=0x960] [pid=303926] [2020-03-11 14:21:38 +1300]
               | Falcon v0.35.5 taking flight! Using Async::Container::Threaded {:count=>8}.
               | - Binding to: #<Falcon::Endpoint https://localhost:9292/ {}>
               | - To terminate: Ctrl-C or kill 303926
               | - To reload configuration: kill -HUP 303926
 0.26s     info: Async::Container::Thread::Instance [oid=0xba4] [pid=303926] [2020-03-11 14:21:38 +1300]
               | - Per-process status: kill -USR1 303926
config.ru:6: warning: already initialized constant DB
config.ru:6: warning: previous definition of DB was here
config.ru:6: warning: already initialized constant DB
config.ru:6: warning: previous definition of DB was here
config.ru:6: warning: already initialized constant DB
config.ru:6: warning: previous definition of DB was here
config.ru:6: warning: already initialized constant DB
config.ru:6: warning: previous definition of DB was here
config.ru:6: warning: already initialized constant DB
config.ru:6: warning: previous definition of DB was here
config.ru:6: warning: already initialized constant DB
config.ru:6: warning: previous definition of DB was here
config.ru:6: warning: already initialized constant DB
config.ru:6: warning: previous definition of DB was here
```

**Comment by ioquatix** [comment]
https://github.com/socketry/falcon/tree/master/examples/sequel is the code + a sample database just in case you want to try it out / break it.

**Comment by jeremyevans** [comment]
> > Even if you were able to get app-per-thread working for the app itself, you would have to ensure all everything inside the application was duplicated, which again, will be true for very few apps.
> 
> That's not true. If you want to use a global shared cache, you need to be sure it's thread safe, and that's okay.

Isn't that getting back to "the app needs to be thread-safe"? My point is you cannot take an arbitrary single-threaded app, run it app-per-thread, and have the result be thread safe.  Ergo, app-per-thread is not a workable default approach, and not one Rack should endorse.  Of course, Rack servers (e.g. falcon) are free to implement the approach, but I don't think Rack should be recommending it.

> > Rails.application is a singleton: https://github.com/rails/rails/blob/master/railties/lib/rails.rb#L39-L41. If you have multiple separate Rails applications in the same process, Rails.application cannot be correct for both.
> 
> That's really an issue with the design of Rails, not Rack, though. I mean, sure, it might not work (but it seems to be fine in my testing). As long as the application is thread safe, I don't think it's an issue that `config.ru` returns an app that points at the same global application instance,

Of course, is the application is thread safe, there is no reason to use app-per-thread, is there?

> Also, I tried your example, and it seems to work fine. Althought I get what you are trying to show.

If it doesn't at least raise an error, you aren't issuing requests to it (the example didn't create the table it uses, so it should definitely raise an exception if issued a request).  Whether it segfaults depends on the behavior of the ruby-sqlite3 driver when multiple threads issue queries on the same connection.

In any case, it just serves as an example, so as long as you get what I'm trying to show, that's what is important.

**Comment by ioquatix** [comment]
> Isn't that getting back to "the app needs to be thread-safe"?

No, it's "If you want your app to run correctly don't use global mutable state." and "If you decide to use global mutable state, make sure you put a lock around it.". Unfortunately even in Rack, we have an example of code which forgets to use a mutex.

So, roughly speaking, there are two models:

1. One app per server thread/process.
2. One app for all threads/processes.

To me, the option 1 is safer and a more straight forward model. We have a ton of evidence that shows that the 2nd option is also very hard to implement correctly.

> If it doesn't at least raise an error, you aren't issuing requests to it (the example didn't create the table it uses, so it should definitely raise an exception if issued a request). Whether it segfaults depends on the behavior of the ruby-sqlite3 driver when multiple threads issue queries on the same connection.

As per the linked files, I created the table and I checked the output of the request with curl, it seems to work fine. The fact that it may or may not segfault is more fuel for the "thread safety is hard" fire. Because people write code and they don't know if it's thread safe or not. It *seems* to work, ergo, it's thread safe.

> My point is you cannot take an arbitrary single-threaded app, run it app-per-thread, and have the result be thread safe.

My point is, I believe we should document this as a requirement, e.g.:

> Ideally, application instances are thread-safe. That is, invoking `#call(env)` multiple times from different threads should not produce undesirable behaviour. In addition, application instances should present an interface that is thread-safe, even when multiple instances are created. The behaviour and public state of one instance should not visibly affect the behaviour of another instance.
>
> If your application is not thread-safe, ensure that you use a multi-process application server. Some servers support an isolated multi-thread model, however as your application is running in the same process, you may need to be careful with global mutable state.

I'm open to editing the above, to something we all agree on, but I do think it should be clearly documented. We should clearly state the models under which the code can run, so that servers can implement consistent environments for applications. Otherwise, we just end up with applications being server-specific which is the anti-thesis of rack.

**Comment by jeremyevans** [comment]
I'm OK with some of the language you are proposing.  I would drop the discussion of application instances, as applications can be singletons (e.g. classes).  I would also drop discussion of the isolated multi-thread model.  So just:

> Ideally, rack applications should be thread-safe. That is, invoking `#call(env)` multiple times from different threads should not produce undesirable behaviour.  If your application is not thread-safe, ensure that you use a single-threaded application server.

**Comment by boazsegev** [comment]
@ioquatix 

> So, I personally believe we should document that such a design is ill-formed. 

Just so we're clear - this "ill-formed" design describes **all** of the applications I've written and reviewed so far in all of my years programming in Ruby.

Based on the current Rack specification, where "_A Rack application is a Ruby object (not a class)_", all of the Rack applications and frameworks I've had experience working with define the Application as either a global object, a singleton class or a Model (with `self.call` defined as a Model method, **not** an instance method).

> roughly speaking, there are two models:
>
> 1. One app per server thread/process.
> 2. One app for all threads/processes.

As @jeremyevans pointed out, neither model adds anything to the question of thread safety.

You're focusing only on the APP object while there may be thousands of other places where non-thread-safe code could be written.

Concurrency concerns in all the code that goes into a project is really out of scope for the Rack specification.

P.S.

We code to change the world - change requires mutability and mutability requires a way to manage race conditions. It's unavoidable, really, even in languages that try to hide it from the developers (such as Clojure, Lisp or Haskal).

All the applications I know use a state machine somewhere - either indirectly (i.e., state stored in an external database / service) or directly (i.e., caching database results, managing database connection pools or adding WebSockets to a subscription array).

Using per-thread object instances to avoid race conditions might solve the issue for some use-cases, but experience shows it doesn't solve the issue for all use cases.

At the end of the day, developers need to know what they're doing or use code (gems) written by people who do.

IMHO, Rack doesn't need to teach developers how to code. I think that anything more than the suggested short friendly reminder (not to use multi-threaded servers unless your application is thread-safe) is excessive.

**Comment by ioquatix** [comment]
> I've had experience working with define the Application as either a global object, a singleton class or a Model (with self.call defined as a Model method, not an instance method).

I have a strong aversion to global mutable state because I've seen it cause so many problems and so many bad designs. Maybe your design is great. But I'd prefer a framework that uses instances of objects to encapsulate state. Isolation is a *very* important property for understanding the behaviour of code.

> As @jeremyevans pointed out, neither model adds anything to the question of thread safety.

It's trivial to show this using an example.

```ruby
class App
  def initialize
    @count = 0
  end
  
  def call(env)
    @count = count + 1
    return [200, [], [@count.to_s]
  end
end

run App.new
```

Using a shared instance of `App` is not thread safe, unless you make one instance per thread. The other option is you expect users to write thread-safe middleware.

> Just so we're clear - this "ill-formed" design describes all of the applications I've written and reviewed so far in all of my years programming in Ruby.

Part of this is the fault of `Rack::Builder` making it non-trivial to do any kind of dependency injection. Ruby also makes this too easy, but it may change with the introduction of Guilds.

> You're focusing only on the APP object while there may be thousands of other places where non-thread-safe code could be written.

I spent a month preparing [this talk](https://www.youtube.com/watch?v=qX_FRa43r4k) where I identify a handful of "other places" where "non-thread-safe code could be writtten". So you are preaching to the choir here.

> Using per-thread object instances to avoid race conditions might solve the issue for some use-cases, but experience shows it doesn't solve the issue for all use cases.

It does solve the issue for many situations, as shown above. I'm glad we agree on this. I don't believe users (including myself) are capable of writing bug-free multi-threading code, especially when that code is as non-trivial as a web application.

This not just about a short friendly reminder, or teaching people how to code, but documenting the expected thread safety requirements of rack middleware, and how servers are expected to behave w.r.t. middleware. However, if we are willing to leave it as a wild west, I'll probably document the expectations on the falcon server README, but I think it's a big loss if we can't come to any consensus here.

**Comment by boazsegev** [comment]
@ioquatix 

> Using a shared instance of App is not thread safe, unless you make one instance per thread. 

When you presented this code:

```ruby
class App
  def initialize
    @count = 0
  end
  
  def call(env)
    @count = @count + 1
    return [200, [], [@count.to_s]
  end
end

run App.new
```

What was the developer's intention? Did they desire each thread to provide a different counter or did they desire the counter to be an application wide state?

This ambiguity isn't something Rack (or servers) need to solve for the developer. This is ambiguous code that the developer needs to clarify by either using thread storage or a Mutex.

> This not just about a short friendly reminder, or teaching people how to code, but documenting the expected thread safety requirements of rack middleware, and how servers are expected to behave w.r.t. middleware. 

I believe that choosing a multi-threaded server is, inherently, a choice that defines the application's requirements (be thread-safe).

However, since this choice might be transparent when a multi-threaded server is used as a default server (i.e., Puma in Rails), it's probably worth to mention that applications should be thread-safe.

For this reason, I believe that [the format @Jeremy suggested](https://github.com/rack/rack/issues/1617#issuecomment-597461825) would be best.

> I think it's a big loss if we can't come to any consensus here.

I think we can reach a consensus on [this format](https://github.com/rack/rack/issues/1617#issuecomment-597461825).

However, IMHO, any additional requirements are out of scope for Rack. I really think we shouldn't force servers to instantiate a new Application instance per thread.

If people don't want threads, they can use processes instead, or use a different server.

**Comment by ioquatix** [comment]
> What was the developer's intention? Did they desire each thread to provide a different counter or did they desire the counter to be an application wide state?

(As a funny/tragic aside: If you run the above code on Rubinius, at one point, it was actually possible for the instance variables to disappear: assigning to ivars caused an internal race condition.)

I like framing it from the POV of the developer.

So, if a developer wrote the above code, the only possible execution model is a single app/process...?

But any modern rack server will typically spin up one or more processes and/or threads and/or fibers. So, in that case, it's actually ill-formed. So, by default, it's broken.

My goal is simply to provide a default position, that such middleware is valid, in the same way it works on a multi-process server, as it does on a multi-thread server. My original assertion being:

>  I think the safest option is to define rack apps as unsafe to invoke from multiple threads.

The point being, can we define a model for thread safety within rack which allows users to write middleware and have it run on both multi-process and multi-thread servers?

> If your application is not thread-safe, ensure that you use a single-threaded application server.

This isn't strictly true, because it *should* be okay to run an application on a multi-threaded server where those threads are isolated.

However, maybe in practice it's not possible. Well, I'd still assert for Rack 3 we have the option to make such guarantees. I'm not saying we shouldn't force servers to do one thing or another, just that, as an example, sharing an app between multiple threads is an *unsafe* default and I think we should document that.

**Comment by ioquatix** [comment]
> I believe that choosing a multi-threaded server is, inherently, a choice that defines the application's requirements (be thread-safe).

This isn't necessarily true. JRuby and TruffleRuby don't support fork, so the model for parallelism doesn't intrinsically relate to the applications requirements (being thread safe).

**Comment by boazsegev** [comment]
@ioquatix ,

> I like framing it from the POV of the developer.
>
> So, if a developer wrote the above code, the only possible execution model is a single app/process...?

Probably. I would assume a developer writing this code would limit the server to a single working process... though, if I were to write that code, I would also limit the server to a single thread ;-)

> But any modern rack server will typically spin up one or more processes and/or threads and/or fibers.

Both Puma and Iodine provide a single thread / single process option (i.e., `iodine -t 1`). Threads and processes are always a **choice** left to the application developer.

I agree that it would be better to highlight the fact that the default choice (for some servers) is to spawn a number of threads and `fork` a number of processes.

I agree that it would be nice to succinctly point that out, [as described here](https://github.com/rack/rack/issues/1617#issuecomment-597461825)

I would also go as far as to add a **suggestion**, such as:

"_It is also recommended that Applications and Middleware avoid mutable instance variables and limit any mutable state storage to the `env` hash object and/or a separate, thread safe, module._"
 
However, I wouldn't **require** any such thing. It's not my place to do so and IMHO it's out of scope for Rack.

> it should be okay to run an application on a multi-threaded server where those threads are isolated.

It's impossible to isolate threads. That's the difference between calling `fork` and calling `pthread_create` (or whatever API your OS provides).

Threads do not include a virtual memory address map, which means that threads always have access to each others memory space. This is the difference between processes and threads and why multi-threaded applications are inherently vulnerable to race conditions.

Even within the Ruby layer, using a per thread application instance won't solve the issue. At most it would delay bugs from being exposed by the subset of developers that neglect to practice encapsulation.

> JRuby and TruffleRuby don't support fork, so the model for parallelism doesn't intrinsically relate to the applications requirements (being thread safe).

The fact that these platforms don't support `fork` does not change the fact that the developer can **choose** either a multi-threaded server or a single threaded server... though it might suggest a poor choice of platform for applications that aren't designed to be thread-safe.

**Comment by jeremyevans** [comment]
> So, after looking through the code changes required for #1618, I found that the session code is _still_ racey:
> 
> https://github.com/rack/rack/blob/6196377654b7ff7ce7abaecea62bb285d77d53aa/lib/rack/session/pool.rb#L39-L44
> 
> It:
> 
> * Assumes that there is a shared `@pool` which isn't always true (multi-process, multi-thread with unique app, guild, etc).
> * `generate_sid` is invoked from outside of the mutex context, creating race conditions.
> 
> So, not even rack itself is thread safe.

I looked into this issue, since I think it is better to fix bugs then to use them to further a point. :)

In `rack/session/pool.rb`, `generate_sid` is only called inside `with_lock` blocks, which do use the mutex.  So internal use in that file is thread safe.

In `rack/session/abstract.id.rb`, it is only called inside `commit_session`, and then only when the `:drop` or `:renew` options are used and the session does not already have an existing id.  I think that case is not thread safe.  I've added a PR to fix things: #1620 .

**Comment by nateberkopec** [comment]
> Even within the Ruby layer, using a per thread application instance won't solve the issue. At most it would delay bugs from being exposed by the subset of developers that neglect to practice encapsulation.

Right. Other than rack middleware, the other common problem I see reported to Puma is people putting database connections (not pools) in globals or constants (REDIS = ...). The model proposed doesn't fix that. The only other place we ever see them is in Rack middleware, which I'm looking at providing some additional stuff in Puma to maybe make that harder to mess up.

So just to sum up for me: 

> Do nothing. Apps which are not thread safe continue to crash. Bad for users.

Works for me. Not up to Rack to tell app servers whats good or not for their users.

> Servers should default to instantiating one app per thread.

Can't guarantee thread-safety, and I'm still skeptical this actually works when apps are singletons but w/e. 

> Document that rack apps must be thread safe.

Agreed that this imposes overhead on singlethreaded servers. 

> Allow servers to advertise thread safety in a standard way 

🤷‍♂ 

> Allow applications to communicate to the server that they are thread safe or not.

No one ever really knows the answer to this question, and people who think their app is not thread-safe don't use threaded modes in app servers.

And, as stated, I disagree fundamentally with the premise that "the point of rack is to bridge web applications and servers". The point of Rack is define a functional interface, not tell application servers how to run that function.

**Comment by ioquatix** [comment]
Thanks for the great discussion, I know it consumes a lot of mental effort, so it is appreciated.

So, I'm looking at how I should implement this in falcon, and there are two primary concerns which I think impact the implementation:

- Graceful reloading, and
- Existing behaviour of shared state.

There are ultimately two models by which apps can be loaded:

```ruby
app = load_config_ru
n.times{make_worker(app)}
```

```ruby
n.times{make_worker(load_config_ru)}
```

(where `make_worker` is either `Thread.new` or `fork`).

In the case of a multi-thread server, the consensus seems to be that the first model is more typical in multi-threaded servers. Thanks for everyone's input on this thus far.

However, in multi-process servers, what is the typical implementation? Is it the same? Load the app in the parent process and fork?

From my POV, I'm considering how to handle graceful reloads, and obviously there are different ways to approach it. The main difference is the complexity of the implementation (e.g. multiple layers to try and manage global state).

I took a look in Rails and they do have some code to detect `fork` it seems, so I wonder how prolific this has become, I also recall seeing this in `redis-rb`, basically invalidating the connection pool on fork or something to that effect.

**Comment by jeremyevans** [comment]
For Unicorn, the default is to load the app separately in each worker after the fork.  You can use the `preload_app` option to load it before forking.  There is also the `worker_exec` option, where after forking, the process execs itself so that it uses a completely separate memory space (making some attacks more difficult or impossible).

Note that the standard multi-process model when not preloading an app is similar to:

```ruby
n.times{fork{load_config_ru}}
```

Basically, the load_config_ru is performed inside the forked copy, not in the parent and then inherited by the fork.

Unicorn has before and after fork hooks, and if you are preloading the app, disconnecting sockets and such before forking is the way to go.  Sequel does not do fork detection by design, so you have to manually disconnect before forking if using Sequel.  Fork detection often doesn't handle all cases, as things like `fork{exit}` can change the state of existing database connections in ways that will break the parent process.

**Comment by ioquatix** [comment]
Thanks for the correction, I've updated my code/examples.

**Comment by boazsegev** [comment]
@ioquatix ,

> there are two primary concerns which I think impact the implementation:
>
> - Graceful reloading, and
> - Existing behaviour of shared state.

This seems somewhat out of scope for both Rack and this issue... but here's my 2¢, from designing iodine.

The best approach really depends on the expected behavior for the "Graceful reloading".

**Supporting "Hot Updates"**

One behavior is closer to "hot application updates", where the application code is allowed to change between reload operations.

Under this design, the root process will never load the application, only worker processes load the application instance (allowing re-spawned child processes to load the new application version).

There are cons and pros for this approach. For example, this approach consumes more resources, as every child process will need to load the application separately and many immutable constants will have separate copies in memory.

**Application Preloading**

Another behavior is simply "cleanup and leaks and history", where the root process will load the application but never processes any requests.

This allows any child process to inherit a "clean slate", leveraging the OS memory model of "copy on write" and minimizing resource usage.

This also improves the reloading performance.

However, besides ignoring live code updates (the application isn't loaded again), this adds complexity to connection pools and pipes (such as used by databases - ActiveRecord / Redis / etc'). These "socket" based resources might end up being shared across process boundaries, where process bound Mutexes and locks might not protect them.

Obviously, these two models are mutually exclusive.

**Comment by ioquatix** [comment]
Just FYI, you absolutely can support hot updates with application preloading, you just need a more complex model:

```
Root process - bind socket
  Cluster process - load app
    Worker process - handle requests
```

To hot reload you spin up a new cluster process. All workers share the same clean cluster preloaded app.

However, making this work on JRuby/TruffleRuby... is a bit more complicated.

**Comment by jeremyevans** [comment]
@ioquatix It seems like most people who responded are against Rack defining an elaborate thread safety specification.  I still think it is best if we limit any mention of thread safety to something like the language I provided in https://github.com/rack/rack/issues/1617#issuecomment-597461825.  Would you be OK with closing this, and if you want a small mention of thread safety in the spec, submitting it as a pull request?

**Comment by ioquatix** [comment]
@jeremyevans I'm not sure where I said we should define an elaborate thread safety specification.

What I still believe is required is some kind of well defined model for the contests under with a `config.ru` can be loaded and the nature in which one or more requests can invoke the `#call` method of a middleware. However, my feeling is that adjustments to the builder interface can be used to communicate that, so I'm happy with it being a technical problem and solving it in a PR to `Rack::Builder`.
